src/audio/copier/copier.c

```diff
@@ -465,20 +556,27 @@ static int create_dai(struct comp_dev *parent_dev, struct copier_data *cd,
 		return -EINVAL;
 	}
 
-	for (i = 0; i < dai_count; i++) {
-		int ret;
-
-		dai.dai_index = dai_index[i];
-		ret = init_dai(parent_dev, drv, config, copier, pipeline, &dai, type, i);
+	if (dai_count == 1) {
+		/* optimized single dai removal */
+		dai.dai_index = dai_index[0];
+		ret = init_dai_single(parent_dev, drv, config, copier, pipeline, &dai, type, 0);
 		if (ret) {
-			comp_err(parent_dev, "failed to create dai");
+			comp_err(parent_dev, "failed to create single endpoint dai");
 			return ret;
 		}
-	}
+	} else {
+		/* multi dai creation and handling */
+		for (i = 0; i < dai_count; i++) {
+			dai.dai_index = dai_index[i];
+			ret = init_dai(parent_dev, drv, config, copier, pipeline, &dai, type, i);
```

Comment:
what you do effectively in this and other patches is
1. you extract a common part from one of dai-zephyr.c functions that you want to expose as:
```
dai_something()
{
	before_something();
	dai_zephyr_something();
	after_something();
}
```
this can be a separate commit, it introduces no changes and shouldn't break anything, right?
2. then in copier.c you do
```
-	obj->drv->ops.something();
+	before_something();
+	dai_zephyr_something();
+	after_something();
```
where you can add `before_something()` and `after_something()` as needed as static functions in copier.c too with comments, where contents comes from. These two patches should be small enough and easy to review and cause no run-time changes.
3. then you can do any further changes that actually represent the essence of the modifications to the code that you want to make.